tests/test_actual_langservers.py

```diff
@@ -0,0 +1,551 @@
+
+    # threaded
+    def _read_loop(self):
+        try:
+            while self._pout:
+                data = self._pout.read(1)
+
+                if data == b"":
+                    break
+
+                self._read_q.put(data)
+        except Exception as ex:
+            self.exception = ex
+        self._send_q.put_nowait(None)  # stop send loop
+
+    # threaded
+    def _send_loop(self):
+        try:
+            while self._pin:
+                buf = self._send_q.get()
+
+                if buf is None:
+                    break
+
+                # print(f"\nsending: {buf}\n")
+
+                self._pin.write(buf)
+                self._pin.flush()
+        except Exception as ex:
+            self.exception = ex
+
+    def _queue_data_to_send(self):
+        send_buf = self.lsp_client.send()
+        if send_buf:
+            self._send_q.put(send_buf)
+
+    def _read_data_received(self):
+        while not self._read_q.empty():
+            data = self._read_q.get()
+            events = self.lsp_client.recv(data)
+            for ev in events:
+                self.msgs.append(ev)
+                self._try_default_reply(ev)
+
+    def _try_default_reply(self, msg):
+        if isinstance(
+            msg,
+            (
+                lsp.ShowMessageRequest,
+                lsp.WorkDoneProgressCreate,
+                lsp.RegisterCapabilityRequest,
+                lsp.ConfigurationRequest,
+            ),
+        ):
+            msg.reply()
+
+        elif isinstance(msg, lsp.WorkspaceFolders):
+            msg.reply([lsp.WorkspaceFolder(uri=self.root_uri, name="Root")])
+
+    #        else:
+    #            print(f"Can't autoreply: {type(msg)}")
+
+    def wait_for_message_of_type(self, type_, timeout=5):
+        end_time = time.time() + timeout
+        while time.time() < end_time:
+            self._queue_data_to_send()
+            self._read_data_received()
+
+            # raise thread's exception if have any
+            if self.exception:
+                raise self.exception
+
+            for msg in self.msgs:
+                if isinstance(msg, type_):
+                    self.msgs.remove(msg)
+                    return msg
+
+            time.sleep(0.2)
+
+        raise Exception(
+            f"Didn't receive {type_} in time; have: " + pprint.pformat(self.msgs)
+        )
+
+    def exit_cleanly(self):
+        # Not necessarily error, gopls sends logging messages for example
+        #        if self.msgs:
+        #            print(
+        #                "* unprocessed messages: " + pprint.pformat(self.msgs)
+        #            )
+
+        assert self.lsp_client.is_initialized
+        self.lsp_client.shutdown()
+        self.wait_for_message_of_type(lsp.Shutdown)
+        self.lsp_client.exit()
+        self._queue_data_to_send()
+        self._read_data_received()
+
+    def do_method(self, text, file_uri, method, response_type=None):
+        def doc_pos():  # SKIP
```

Comment:
@halfbrained What doees this comment mean?

Reply:
It's for my editor :) no other purpose